src/machine/core.cpp

```diff
@@ -0,0 +1,1000 @@
+        if (!dt.bjr_req_rs) {
+            regs->pc_abs_jmp_28(dt.inst.address() << 2);
+            emit fetch_jump_value(true);
+            emit fetch_jump_reg_value(false);
+        } else {
+            regs->pc_abs_jmp(Address(dt.val_rs.as_u32()));
+            emit fetch_jump_value(false);
+            emit fetch_jump_reg_value(true);
+        }
+        emit fetch_branch_value(false);
+        return true;
+    }
+
+    if (dt.branch) {
+        if (dt.bjr_req_rt) {
+            branch = dt.val_rs.as_u32() == dt.val_rt.as_u32();
+        } else if (!dt.bgt_blez) {
+            branch = dt.val_rs.as_i32() < 0;
+        } else {
+            branch = dt.val_rs.as_i32() <= 0;
+        }
+
+        if (dt.bj_not) {
+            branch = !branch;
+        }
+    }
+
+    emit fetch_jump_value(false);
+    emit fetch_jump_reg_value(false);
+    emit fetch_branch_value(branch);
+
+    if (branch) {
+        int32_t rel_offset = dt.inst.immediate() << 2;
+        if (rel_offset & (1 << 17)) {
+            rel_offset -= 1 << 18;
+        }
+        regs->pc_abs_jmp(dt.inst_addr + rel_offset + 4);
+    } else {
+        regs->pc_inc();
+    }
+    return branch;
+}
+
+void Core::dtFetchInit(struct dtFetch &dt) {
+    dt.inst = Instruction(0x00);
+    dt.excause = EXCAUSE_NONE;
+    dt.in_delay_slot = false;
+    dt.is_valid = false;
+}
+
+void Core::dtDecodeInit(struct dtDecode &dt) {
+    dt.inst = Instruction(0x00);
+    dt.memread = false;
+    dt.memwrite = false;
+    dt.alusrc = false;
+    dt.regd = false;
+    dt.regwrite = false;
+    dt.bjr_req_rs = false; // requires rs for beq, bne, blez, bgtz, jr nad
+                           // jalr
+    dt.bjr_req_rt = false; // requires rt for beq, bne
+    dt.jump = false;
+    dt.bj_not = false;
+    dt.bgt_blez = false;
+    dt.nb_skip_ds = false;
+    dt.forward_m_d_rs = false;
+    dt.forward_m_d_rt = false;
+    dt.aluop = ALU_OP_SLL;
+    dt.memctl = AC_NONE;
+    dt.num_rs = 0;
+    dt.num_rt = 0;
+    dt.num_rd = 0;
+    dt.val_rs = 0;
+    dt.val_rt = 0;
+    dt.rwrite = 0;
+    dt.immediate_val = 0;
+    dt.ff_rs = FORWARD_NONE;
+    dt.ff_rt = FORWARD_NONE;
+    dt.excause = EXCAUSE_NONE;
+    dt.in_delay_slot = false;
+    dt.stall = false;
+    dt.stop_if = false;
+    dt.is_valid = false;
+}
+
+void Core::dtExecuteInit(struct dtExecute &dt) {
+    dt.inst = Instruction(0x00);
+    dt.memread = false;
+    dt.memwrite = false;
+    dt.regwrite = false;
+    dt.memctl = AC_NONE;
+    dt.val_rt = 0;
+    dt.rwrite = 0;
+    dt.alu_val = 0;
+    dt.excause = EXCAUSE_NONE;
+    dt.in_delay_slot = false;
+    dt.stop_if = false;
+    dt.is_valid = false;
+}
+
+void Core::dtMemoryInit(struct dtMemory &dt) {
+    dt.inst = Instruction(0x00);
+    dt.memtoreg = false;
+    dt.regwrite = false;
+    dt.rwrite = false;
+    dt.towrite_val = 0;
+    dt.mem_addr = 0x0_addr;
+    dt.excause = EXCAUSE_NONE;
+    dt.in_delay_slot = false;
+    dt.stop_if = false;
+    dt.is_valid = false;
+}
```

Comment:
Move to default constructors. Initialize with RV NOP. Avoid existence of uninitialized registers.